tokio-net/src/driver/sharded_slab/mod.rs

```diff
@@ -0,0 +1,317 @@
+    // │             │ │    │XXXXXXXX│ │
+    // │ local_free──┼─┘    ├────────┤ │
+    // │ global_free─┼─┐    │        │◀┘
+    // ├─────────────┤ └───▶│  next──┼─┐
+    // │   page 3    │      ├────────┤ │
+    // └─────────────┘      │XXXXXXXX│ │
+    //       ...            ├────────┤ │
+    // ┌─────────────┐      │XXXXXXXX│ │
+    // │ page n      │      ├────────┤ │
+    // └─────────────┘      │        │◀┘
+    //                      │  next──┼───▶
+    //                      ├────────┤
+    //                      │XXXXXXXX│
+    //                      └────────┘
+    //                         ...
+    pages: Vec<Page<T>>,
+}
+
+impl<T> Slab<T> {
+    pub(crate) const MAX_BIT: usize = slot::Generation::LEN + slot::Generation::SHIFT;
+
+    /// Returns a new slab.
+    pub(crate) fn new() -> Self {
+        let mut shards = Vec::with_capacity(MAX_THREADS);
+        let mut idx = 0;
+        shards.resize_with(MAX_THREADS, || {
+            let shard = Shard::new(idx);
+            idx += 1;
+            CausalCell::new(shard)
+        });
+        Slab {
+            shards: shards.into_boxed_slice(),
+        }
+    }
+
+    /// Inserts a value into the slab, returning a key that can be used to
+    /// access it.
+    ///
+    /// If this function returns `None`, then the shard for the current thread
+    /// is full and no items can be added until some are removed, or the maximum
+    /// number of shards has been reached.
+    pub(crate) fn insert(&self, value: T) -> Option<usize> {
+        let tid = Tid::current();
+        self.shards[tid.as_usize()]
+            .with_mut(|shard| unsafe {
+                // we are guaranteed to only mutate the shard while on its thread.
+                (*shard).insert(value)
+            })
+            .map(|idx| tid.pack(idx))
+    }
+
+    /// Removes the value associated with the given key from the slab, returning
+    /// it.
+    ///
+    /// If the slab does not contain a value for that key, `None` is returned
+    /// instead.
+    pub(crate) fn remove(&self, idx: usize) -> Option<T> {
+        let tid = Tid::from_packed(idx);
+        if tid.is_current() {
+            self.shards[tid.as_usize()].with_mut(|shard| unsafe {
+                // only called if this is the current shard
+                (*shard).remove_local(idx)
+            })
+        } else {
+            self.shards[tid.as_usize()].with(|shard| unsafe { (*shard).remove_remote(idx) })
+        }
+    }
+
+    /// Return a reference to the value associated with the given key.
+    ///
+    /// If the slab does not contain a value for the given key, `None` is
+    /// returned instead.
+    pub(crate) fn get(&self, key: usize) -> Option<&T> {
+        let tid = Tid::from_packed(key);
+        self.shards
+            .get(tid.as_usize())?
+            .with(|shard| unsafe { (*shard).get(key) })
+    }
+
+    /// Returns the number of items currently stored in the slab.
+    pub(crate) fn len(&self) -> usize {
```

Comment:
Could we remove this from slab and add it directly to the driver's struct? Add a counter incremented in `add_source` and decremented in `drop_source`. Let's use `SeqCst` for ordering. I expect the counter to go away in the near future. It would be easier to manage keeping it separate from the slab.